dmoj/commands/validate.py

```diff
@@ -0,0 +1,182 @@
+        self.arg_parser.add_argument('problem_ids', nargs='+', help='ids of problems to test')
+
+    def execute(self, line: str) -> int:
+        args = self.arg_parser.parse_args(line)
+
+        problem_ids = args.problem_ids
+        supported_problems = set(get_supported_problems())
+
+        unknown_problems = ', '.join(f"'{i}'" for i in problem_ids if i not in supported_problems)
+        if unknown_problems:
+            raise InvalidCommandException(f'unknown problem(s) {unknown_problems}')
+        total_fails = 0
+        for problem_id in problem_ids:
+            if not self.validate_problem(problem_id):
+                print_ansi(f'Problem #ansi[{problem_id}](cyan|bold) #ansi[failed](red|bold).')
+                total_fails += 1
+            else:
+                print_ansi(f'Problem #ansi[{problem_id}](cyan|bold) passed with flying colours.')
+            print()
+
+        print()
+        print('Test complete.')
+        if total_fails:
+            print_ansi(f'#ansi[A total of {total_fails} problem(s) have invalid input](red|bold)')
+        else:
+            print_ansi('#ansi[All problems validated.](green|bold)')
+
+        return total_fails
+
+    def validate_problem(self, problem_id: str) -> bool:
+        print_ansi(f'Validating problem #ansi[{problem_id}](cyan|bold)...')
+
+        problem_root = get_problem_root(problem_id)
+        assert problem_root is not None
+        config = ProblemConfig(ProblemDataManager(problem_root))
+
+        if problem_root is None:
+            print_ansi(f'\t#ansi[Skipped](magenta|bold) - Problem [{problem_id}](cyan|bold) not found')
+            return True
+
+        if not config or 'validator' not in config:
+            print_ansi('\t#ansi[Skipped](magenta|bold) - No validator found')
+            return True
+
+        validator_config = config['validator']
+        language = validator_config['language']
+        if language not in all_executors:
+            print_ansi('\t\t#ansi[Skipped](magenta|bold) - Language not supported')
+            return True
+        time_limit = validator_config.get('time', env.validator_time_limit)
+        memory_limit = validator_config.get('memory', env.validator_memory_limit)
+        compiler_time_limit = validator_config.get('compiler_time_limit', env.validator_compiler_time_limit)
+        read_feedback_from = validator_config.get('read_feedback_from', 'stderr')
+
+        if read_feedback_from not in ('stdout', 'stderr'):
+            print_ansi('\t\t#ansi[Failed](red|bold) - Feedback option should be (stdout, stderr)')
+            return False
+
+        if isinstance(validator_config.source, str):
+            filenames = [validator_config.source]
+        elif isinstance(validator_config.source.unwrap(), list):
+            filenames = list(validator_config.source.unwrap())
+        else:
+            print_ansi('\t#ansi[Skipped](magenta|bold) - No validator found')
+            return True
```

Comment:
If a validator config exists but no sources are found, maybe this isn't "skip"? Maybe this should be a failure?